### PR TITLE
ios: marked deleted chat items, full deletion preference; android: types

### PR DESCRIPTION
**Pull request by jr-simplex**

### CODE REVIEW

File: apps/ios/Shared/Views/Chat/ChatItem/CIVoiceView.swift [hunk 1]
```diff
@@ -223,7 +223,7 @@ struct CIVoiceView_Previews: PreviewProvider {
             quotedItem: nil,
             file: nil
         )
-        Group {
+        return Group {
```
Comment (epoberezkin) [inline apps/ios/Shared/Views/Chat/ChatItem/CIVoiceView.swift:226]:
```suggestion
        Group {
```

File: apps/android/app/src/main/java/chat/simplex/app/views/chat/item/FramedItemView.kt [hunk 2]
```diff
@@ -131,15 +132,15 @@ fun FramedItemView(
       Column(Modifier.width(IntrinsicSize.Max)) {
         PriorityLayout(Modifier, CHAT_IMAGE_LAYOUT_ID) {
           ci.quotedItem?.let { ciQuoteView(it) }
-          if (ci.file == null && ci.formattedText == null && isShortEmoji(ci.content.text)) {
```
Comment (epoberezkin) [inline apps/android/app/src/main/java/chat/simplex/app/views/chat/item/FramedItemView.kt:134]: is it always the same?

Comment (epoberezkin) [reply apps/android/app/src/main/java/chat/simplex/app/views/chat/item/FramedItemView.kt:134]: I am not sure why this change is needed, and I am actually not sure it is the same.

File: apps/android/app/src/main/java/chat/simplex/app/views/chat/item/FramedItemView.kt [hunk 3]
```diff
@@ -131,15 +132,15 @@ fun FramedItemView(
...
+          if (ci.file == null && ci.formattedText == null && isShortEmoji(ci.text)) {
             Box(Modifier.padding(vertical = 6.dp, horizontal = 12.dp)) {
               Column(
                 Modifier
                   .padding(bottom = 2.dp)
                   .fillMaxWidth(),
                 horizontalAlignment = Alignment.CenterHorizontally
               ) {
-                EmojiText(ci.content.text)
+                EmojiText(ci.text)
```
Comment (epoberezkin) [inline apps/android/app/src/main/java/chat/simplex/app/views/chat/item/FramedItemView.kt:143]: same here

File: apps/android/app/src/main/java/chat/simplex/app/model/SimpleXAPI.kt [hunk 4]
```diff
@@ -448,9 +448,9 @@ open class ChatController(var ctrl: ChatCtrl?, val ntfManager: NtfManager, val a
     return null
   }
 
-  suspend fun apiDeleteChatItem(type: ChatType, id: Long, itemId: Long, mode: CIDeleteMode): AChatItem? {
+  suspend fun apiDeleteChatItem(type: ChatType, id: Long, itemId: Long, mode: CIDeleteMode): Pair<ChatItem, ChatItem?>? {
     val r = sendCmd(CC.ApiDeleteChatItem(type, id, itemId, mode))
-    if (r is CR.ChatItemDeleted) return r.toChatItem
+    if (r is CR.ChatItemDeleted) return Pair(r.deletedChatItem.chatItem, r.toChatItem?.chatItem)
```
Comment (epoberezkin) [inline apps/android/app/src/main/java/chat/simplex/app/model/SimpleXAPI.kt:453]: return CR.ChatItemDeleted

File: apps/android/app/src/main/java/chat/simplex/app/model/SimpleXAPI.kt [hunk 5]
```diff
@@ -1082,14 +1082,11 @@ open class ChatController(var ctrl: ChatCtrl?, val ntfManager: NtfManager, val a
       is CR.ChatItemUpdated ->
         chatItemSimpleUpdate(r.chatItem)
       is CR.ChatItemDeleted -> {
-        val cInfo = r.toChatItem.chatInfo
-        val cItem = r.toChatItem.chatItem
-        if (cItem.meta.itemDeleted) {
-          chatModel.removeChatItem(cInfo, cItem)
+        AudioPlayer.stop(r.deletedChatItem.chatItem)
+        if (r.toChatItem != null) {
```
Comment (epoberezkin) [inline apps/android/app/src/main/java/chat/simplex/app/model/SimpleXAPI.kt:1086]: flip if

File: apps/android/app/src/main/java/chat/simplex/app/model/SimpleXAPI.kt [hunk 6]
```diff
@@ -2045,6 +2045,12 @@ data class FeatureEnabled(
 sealed class ContactUserPref {
   @Serializable @SerialName("contact") data class Contact(val preference: ChatPreference): ContactUserPref() // contact override is set
   @Serializable @SerialName("user") data class User(val preference: ChatPreference): ContactUserPref() // global user default is used
+
+  val pref: ChatPreference
+    get() = when(this) {
+      is Contact -> preference
+      is User -> preference
+    }
```
Comment (epoberezkin) [inline apps/android/app/src/main/java/chat/simplex/app/model/SimpleXAPI.kt:2053]: Possibly, you can just use .preference because it is present in both subclasses of sealed class.

If not, it should be defined as an abstract property on superclass level and overridden in subclasses, there is no need to have this computed property at all.

File: apps/android/app/src/main/java/chat/simplex/app/model/SimpleXAPI.kt [hunk 7]
```diff
@@ -2279,7 +2285,10 @@ data class GroupPreferences(
 @Serializable
 data class GroupPreference(
   val enable: GroupFeatureEnabled
-)
+) {
+  val on: Boolean
+    get() = enable == GroupFeatureEnabled.ON
```
Comment (epoberezkin) [inline apps/android/app/src/main/java/chat/simplex/app/model/SimpleXAPI.kt:2290]:
```suggestion
  val on: Boolean get() = enable == GroupFeatureEnabled.ON
```

File: apps/android/app/src/main/java/chat/simplex/app/views/chat/ChatView.kt [hunk 8]
```diff
@@ -166,13 +166,20 @@ fun ChatView(chatId: String, chatModel: ChatModel, onComposed: () -> Unit) {
...
+          val r = chatModel.controller.apiDeleteChatItem(
             type = cInfo.chatType,
             id = cInfo.apiId,
             itemId = itemId,
             mode = mode
           )
-          if (toItem != null) chatModel.removeChatItem(cInfo, toItem.chatItem)
+          if (r != null) {
+            val (deletedItem, toItem) = r
+            if (toItem != null) {
```
Comment (epoberezkin) [inline apps/android/app/src/main/java/chat/simplex/app/views/chat/ChatView.kt:177]: flip if

File: apps/android/app/src/main/java/chat/simplex/app/views/chat/ContextItemView.kt [hunk 9]
```diff
@@ -50,7 +50,7 @@ fun ContextItemView(
         tint = HighOrLowlight,
       )
       MarkdownText(
-        contextItem.text, contextItem.formattedText,
+        contextItem.text_, contextItem.formattedText_,
```
Comment (epoberezkin) [inline apps/android/app/src/main/java/chat/simplex/app/views/chat/ContextItemView.kt:53]: this will go away?

File: apps/android/app/src/main/java/chat/simplex/app/model/NtfManager.kt [hunk 10]
```diff
@@ -218,19 +218,15 @@ class NtfManager(val context: Context, private val appPreferences: AppPreference
...
   private fun hideSecrets(cItem: ChatItem) : String {
-    val md = cItem.formattedText
-    return if (md == null) {
-      if (cItem.content.text != "") {
-        cItem.content.text
-      } else {
-        if (cItem.content.msgContent is MsgContent.MCVoice) generalGetString(R.string.voice_message) else cItem.file?.fileName ?: ""
-      }
-    } else {
+    val md = cItem.formattedText_
```
Comment (epoberezkin) [inline apps/android/app/src/main/java/chat/simplex/app/model/NtfManager.kt:221]: this text_ is completely unnecessary because it simply hides a real condition (meta.itemDeleted) behind == null condition, it's not reducing complexity and only creates an unnecessary indirection.

File: apps/android/app/src/main/java/chat/simplex/app/model/ChatModel.kt [hunk 11]
```diff
@@ -1043,14 +1063,32 @@ data class ChatItem (
   val id: Long get() = meta.itemId
   val timestampText: String get() = meta.timestampText
 
-  val text: String get() =
-    when {
-      content.text == "" && file != null && content.msgContent is MsgContent.MCVoice -> {
-        (content.msgContent as MsgContent.MCVoice).toTextWithDuration(false)
-      }
+  // pair with formattedText
```
Comment (epoberezkin) [inline apps/android/app/src/main/java/chat/simplex/app/model/ChatModel.kt:1066]: what is this comment about?

File: apps/android/app/src/main/java/chat/simplex/app/model/ChatModel.kt [hunk 12]
```diff
@@ -1043,14 +1063,32 @@ data class ChatItem (
...
+    }
+  }
+
+  val formattedText_: List<FormattedText>? get() {
+    if (meta.itemDeleted) {
+      return null
+    } else {
+      return formattedText
+    }
+  }
```
Comment (epoberezkin) [inline apps/android/app/src/main/java/chat/simplex/app/model/ChatModel.kt:1091]: these seem unnecessary / only complicating things

File: apps/android/app/src/main/java/chat/simplex/app/views/chat/ComposeView.kt [hunk 13]
```diff
@@ -457,15 +457,15 @@ fun ComposeView(
 
   fun allowVoiceToContact() {
     val contact = (chat.chatInfo as ChatInfo.Direct?)?.contact ?: return
-    val featuresAllowed = contactUserPrefsToFeaturesAllowed(contact.mergedPreferences)
+    val prefs = contact.mergedPreferences.toPreferences().copy(voice = ChatPreference(allow = FeatureAllowed.YES))
```
Comment (epoberezkin) [inline apps/android/app/src/main/java/chat/simplex/app/views/chat/ComposeView.kt:460]: drive-by refactoring - move to another PR

File: apps/ios/Shared/Views/Chat/ChatItem/FramedItemView.swift [hunk 14]
```diff
@@ -26,79 +26,51 @@ struct FramedItemView: View {
     @State var imgWidth: CGFloat? = nil
     @State var metaColor = Color.secondary
     @State var showFullScreenImage = false
-
+    @Binding var revealed: Bool
+    
     var body: some View {
+        if chatItem.meta.itemDeleted && !revealed {
+            MarkedDeletedItemView(chatItem: chatItem, showMember: showMember)
```
Comment (epoberezkin) [inline apps/ios/Shared/Views/Chat/ChatItem/FramedItemView.swift:33]: if this is how it worked out, it's better to move it one level up, where you decide if you need FramedItemView

File: apps/ios/Shared/Views/Chat/ChatItem/FramedItemView.swift [hunk 15]
```diff
@@ -237,48 +279,77 @@ private struct MetaColorPreferenceKey: PreferenceKey {
...
-func chatItemFrameColor(_ ci: ChatItem, _ colorScheme: ColorScheme) -> Color {
+func chatItemFrameColorMaybeImage(_ ci: ChatItem, _ colorScheme: ColorScheme) -> Color {
     onlyImage(ci)
     ? Color.clear
-    : ci.chatDir.sent
+    : chatItemFrameColor(ci, colorScheme)
+}
+
+func chatItemFrameColor(_ ci: ChatItem, _ colorScheme: ColorScheme) -> Color {
+    ci.chatDir.sent
```
Comment (epoberezkin) [inline apps/ios/Shared/Views/Chat/ChatItem/FramedItemView.swift:294]: why is this change needed?

File: apps/ios/Shared/Views/Chat/ChatItem/FramedItemView.swift [hunk 16]
```diff
@@ -144,21 +190,17 @@ struct FramedItemView: View {
...
-        .frame(minWidth: msgWidth, alignment: .leading)
-        .background(
-            chatItem.chatDir.sent
-            ? (colorScheme == .light ? sentQuoteColorLight : sentQuoteColorDark)
-            : Color(uiColor: .quaternarySystemFill)
-        )
-
+            .overlay(DetermineWidth())
+            .frame(minWidth: msgWidth, alignment: .leading)
+            .background(chatItemFrameContextColor(chatItem, colorScheme))
```
Comment (epoberezkin) [inline apps/ios/Shared/Views/Chat/ChatItem/FramedItemView.swift:195]: what does this change does?

File: apps/ios/Shared/Views/Chat/ChatItem/FramedItemView.swift [hunk 17]
```diff
@@ -237,48 +279,77 @@ private struct MetaColorPreferenceKey: PreferenceKey {
...
+            FramedItemView(chatInfo: ChatInfo.sampleData.direct, chatItem: ChatItem.getSample(2, .directRcv, .now, "hello there too!!! this text has the time on the same line "), revealed: Binding.constant(false))
+            FramedItemView(chatInfo: ChatInfo.sampleData.direct, chatItem: ChatItem.getSample(2, .directRcv, .now, "https://simplex.chat"), revealed: Binding.constant(false))
+            FramedItemView(chatInfo: ChatInfo.sampleData.direct, chatItem: ChatItem.getSample(2, .directRcv, .now, "chaT@simplex.chat"), revealed: Binding.constant(false))
         }
         .previewLayout(.fixed(width: 360, height: 200))
     }
 }
 
-struct FramedItemViewEdited_Previews: PreviewProvider {
+struct FramedItemView_Edited_Previews: PreviewProvider {
```
Comment (epoberezkin) [inline apps/ios/Shared/Views/Chat/ChatItem/FramedItemView.swift:321]: why name change?

File: apps/ios/Shared/Views/Chat/ChatItemView.swift [hunk 18]
```diff
@@ -15,11 +15,40 @@ struct ChatItemView: View {
     var showMember = false
     var maxWidth: CGFloat = .infinity
     @State var scrollProxy: ScrollViewProxy? = nil
+    @Binding var revealed: Bool
+
+    var body: some View {
+        let ci = chatItem
+        if ci.quotedItem == nil && !ci.meta.itemDeleted {
```
Comment (epoberezkin) [inline apps/ios/Shared/Views/Chat/ChatItemView.swift:22]: this is a better place to put condition `deleted & !revealed

File: apps/ios/SimpleXChat/ChatTypes.swift [hunk 19]
```diff
@@ -1374,6 +1383,23 @@ public struct ChatItem: Identifiable, Decodable {
...
+        }
+    }
+
+    public var formattedText_: [FormattedText]? {
+        if meta.itemDeleted {
+            return nil
+        } else {
+            return formattedText
+        }
+    }
```
Comment (epoberezkin) [inline apps/ios/SimpleXChat/ChatTypes.swift:1401]: remove

File: apps/ios/Shared/Views/Chat/ChatView.swift [hunk 20]
```diff
@@ -552,15 +616,19 @@ struct ChatView: View {
...
                         id: chat.chatInfo.apiId,
                         itemId: di.id,
                         mode: mode
                     )
                     DispatchQueue.main.async {
                         deletingItem = nil
-                        let _ = chatModel.removeChatItem(chat.chatInfo, toItem)
+                        if let toItem = toItem {
+                            _ = chatModel.upsertChatItem(chat.chatInfo, toItem)
+                        } else {
```
Comment (epoberezkin) [inline apps/ios/Shared/Views/Chat/ChatView.swift:629]: return

File: apps/android/app/src/main/java/chat/simplex/app/model/NtfManager.kt (hunk 10 above)
Comment (jr-simplex) [reply apps/android/app/src/main/java/chat/simplex/app/model/NtfManager.kt:221]: it only imitates original formatted text - the fact that it could be null

File: apps/ios/Shared/Views/Chat/ChatItem/FramedItemView.swift (hunk 17 above)
Comment (jr-simplex) [reply apps/ios/Shared/Views/Chat/ChatItem/FramedItemView.swift:321]: to separate view name, I've added another PreviewProvider for marked deleted items

File: apps/ios/Shared/Views/Chat/ChatItem/FramedItemView.swift (hunk 16 above)
Comment (jr-simplex) [reply apps/ios/Shared/Views/Chat/ChatItem/FramedItemView.swift:195]: same frame color is used for marked deleted section

File: apps/ios/Shared/Views/Chat/ChatItem/FramedItemView.swift (hunk 15 above)
Comment (jr-simplex) [reply apps/ios/Shared/Views/Chat/ChatItem/FramedItemView.swift:294]: this was affecting marked deleted items

File: apps/ios/Shared/Views/Chat/ChatView.swift (hunk 20 above)
Comment (jr-simplex) [reply apps/ios/Shared/Views/Chat/ChatView.swift:629]: didn't understand
